File: topic1/solutiont11/malloc/code/judge/c/tester.c
#include <assert.h>
#include <errno.h>
#include <float.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "solve.h"
#include "tester.h"
#include "utils.h"

const char *flag =
    "77796755c34b3aada95b7ab2caf33d5ec337d567e58a328cf02c139b29bc650c";

int main(int argc, char **argv) {
    // if (argc != 2) {
    //     return 0;
    // }

    char *bias=(char* )malloc(100 * (1 << 20));
    memset(bias,1,100 * (1 << 20));

    char **g_datafiles = NULL;
    range_t *g_ranges = NULL;
    stats_t *g_stats = NULL;
    speed_t g_speed_params;

    int num_datafiles = 1;
    char datadir[MAXLINE];
    if ((g_datafiles = realloc(g_datafiles, 2 * sizeof(char *))) == NULL)
        unix_error("ERROR: realloc failed in main");
    strcpy(datadir, "./");
    g_datafiles[0] = "sasdsad";
    g_datafiles[1] = NULL;

    init_random_data();

    // if (verbose > 1) {
    //     printf("Testing you malloc\n");
    // }

    g_stats = (stats_t *)calloc((size_t)num_datafiles, sizeof(stats_t));
    if (g_stats == NULL) unix_error("g_stats calloc in main failed");

    run_tests(num_datafiles, datadir, g_datafiles, g_stats, g_ranges,
              &g_speed_params);

    return 0;
}

static void run_tests(int num_tracefiles, const char *tracedir,
                      char **tracefiles, stats_t *mm_stats, range_t *g_ranges,
                      speed_t *speed_para) {
    volatile int i;
    for (i = 0; i < num_tracefiles; ++i) {
        mem_init();
        trace_t *trace;
        trace = read_trace(&mm_stats[i], tracedir, tracefiles[i]);
        strcpy(mm_stats[i].filename, trace->filename);
        mm_stats[i].ops = trace->num_ops;
        if (verbose > 1) printf("Checking for correctness, ");

        mm_stats[i].valid = eval_mm_valid(trace, &g_ranges);
        if (onetime_flag) {
            free_trace(trace);
            return;
        }
        if (mm_stats[i].valid) {
            if (verbose > 1) printf("utilization, ");
            mm_stats[i].util = eval_mm_util(trace, i);

            speed_para->trace = trace;
            speed_para->ranges = g_ranges;

            if (verbose > 1) printf("and time.\n");
            mm_stats[i].secs = timer_gettod(eval_mm_speed, speed_para, 10);
        }

        if (mm_stats[i].valid && error == 0) {
            printf("%s %d %.6f %.6f\n", flag, mm_stats[i].valid,
                   1 - mm_stats[i].util, mm_stats[i].secs);
        } else {
            printf("%s %d %.6f %.6f\n", flag, 0, 1.0, 99999999.0);
        }
        free_trace(trace);
        mem_deinit();
    }
}

static int add_range(range_t **g_ranges, char *lo, int size,
                     const trace_t *trace, int opnum, int index) {
    char *hi = lo + size - 1;
    range_t *p;
    assert(size > 0);

    if (!IS_ALIGNED(lo)) {
        sprintf(msg, "Payload address (%p) not aligned to %d bytes", lo,
                ALIGNMENT);
        // malloc_error(trace, opnum, msg);
        return 0;
    }

    if ((lo < (char *)mem_heap_lo()) || (lo > (char *)mem_heap_hi()) ||
        (hi < (char *)mem_heap_lo()) || (hi > (char *)mem_heap_hi())) {
        sprintf(msg, "Payload (%p:%p) lies outside heap (%p:%p)", lo, hi,
                mem_heap_lo(), mem_heap_hi());
        // malloc_error(trace, opnum, msg);
        return 0;
    }

    for (p = *g_ranges; p != NULL; p = p->next) {
        if ((lo >= p->lo && lo <= p->hi) || (hi >= p->lo && hi <= p->hi)) {
            sprintf(msg, "Payload (%p:%p) overlaps another payload (%p:%p)\n",
                    lo, hi, p->lo, p->hi);
            // malloc_error(trace, opnum, msg);
            return 0;
        }
    }

    if ((p = (range_t *)malloc(sizeof(range_t))) == NULL)
        unix_error("malloc error in add_range");
    p->next = *g_ranges;
    p->lo = lo;
    p->hi = hi;
    p->index = index;
    *g_ranges = p;

    return 1;
}

static void remove_range(range_t **g_ranges, char *lo) {
    range_t *p;
    range_t **prevpp = g_ranges;

    for (p = *g_ranges; p != NULL; p = p->next) {
        if (p->lo == lo) {
            *prevpp = p->next;
            free(p);
            break;
        }
        prevpp = &(p->next);
    }
}

static void clear_ranges(range_t **g_ranges) {
    range_t *p;
    range_t *pnext;

    for (p = *g_ranges; p != NULL; p = pnext) {
        pnext = p->next;
        free(p);
    }
    *g_ranges = NULL;
}

static void init_random_data(void) {
    int len;

    for (len = 0; len < RANDOM_DATA_LEN; ++len) {
        random_data[len] = random();
    }
}

static void randomize_block(trace_t *traces, int index) {
    size_t size;
    size_t i;
    randint_t *block;
    int base;

    traces->block_rand_base[index] = random();

    block = (randint_t *)traces->blocks[index];
    size = traces->block_sizes[index] / sizeof(*block);
    base = traces->block_rand_base[index];

    for (i = 0; i < size; i++) {
        block[i] = random_data[(base + i) % RANDOM_DATA_LEN];
    }
}

static int check_index(const trace_t *trace, int opnum, int index) {
    size_t size;
    size_t i;
    randint_t *block;
    int base;
    int ngarbled = 0;
    int firstgarbled = -1;

    if (index < 0) return 0;

    block = (randint_t *)trace->blocks[index];
    size = trace->block_sizes[index] / sizeof(*block);
    base = trace->block_rand_base[index];

    for (i = 0; i < size; i++) {
        if (block[i] != random_data[(base + i) % RANDOM_DATA_LEN]) {
            if (firstgarbled == -1) firstgarbled = i;
            ngarbled++;
        }
    }
    if (ngarbled != 0) {
        if (verbose > 1) {
            sprintf(msg, "block %d has %d garbled %s%s, starting at byte %zu",
                    index, ngarbled, randint_t_name, ngarbled > 1 ? "s" : "",
                    sizeof(randint_t) * firstgarbled);
            // malloc_error(trace, opnum, msg);
        }
        return 1;
    }
    return 0;
}

static trace_t *read_trace(stats_t *stats, const char *tracedir,
                           const char *filename) {
    // FILE *tracefile;
    trace_t *trace;
    char type[MAXLINE];
    //    char path[MAXLINE];
    unsigned index, size;
    unsigned max_index = 0;
    unsigned op_index;

    if (verbose > 1) {
        printf("Read testfile: %s\n", filename);
    }

    if ((trace = (trace_t *)malloc(sizeof(trace_t))) == NULL)
        unix_error("malloc 1 failed in read_trace");

    strcpy(trace->filename, tracedir);
    strcat(trace->filename, filename);

    // if ((tracefile = fopen(trace->filename, "r")) == NULL) {
    //     sprintf(msg, "Could not open %s in read_trace", trace->filename);
    //     unix_error(msg);
    // }

    fscanf(stdin, "%d", &trace->weight);
    fscanf(stdin, "%d", &trace->num_ids);
    fscanf(stdin, "%d", &trace->num_ops);
    fscanf(stdin, "%d", &trace->ignore_ranges);

    // if (trace->weight < 0 || trace->weight > 1) {
    //     printf("%s: weight can only be in {0, 1}", trace->filename);
    //     exit(1);
    // }
    // if (trace->ignore_ranges != 0 && trace->ignore_ranges != 1) {
    //     printf("%s: flag can only be zero or one", trace->filename);
    //     exit(1);
    // }

    if ((trace->ops =
             (traceop_t *)malloc(trace->num_ops * sizeof(traceop_t))) == NULL)
        unix_error("malloc 2 failed in read_trace");

    if ((trace->blocks = (char **)malloc(trace->num_ids * sizeof(char *))) ==
        NULL)
        unix_error("malloc 3 failed in read_trace");

    if ((trace->block_sizes =
             (size_t *)malloc(trace->num_ids * sizeof(size_t))) == NULL)
        unix_error("malloc 4 failed in read_trace");

    if ((trace->block_rand_base =
             calloc(trace->num_ids, sizeof(*trace->block_rand_base))) == NULL)
        unix_error("malloc 5 failed in read_trace");

    index = 0;
    op_index = 0;
    while (fscanf(stdin, "%s", type) != EOF) {
        switch (type[0]) {
            case 'a':
                fscanf(stdin, "%u %u", &index, &size);
                trace->ops[op_index].type = ALLOC;
                trace->ops[op_index].index = index;
                trace->ops[op_index].size = size;
                max_index = (index > max_index) ? index : max_index;
                break;
            case 'r':
                fscanf(stdin, "%u %u", &index, &size);
                trace->ops[op_index].type = REALLOC;
                trace->ops[op_index].index = index;
                trace->ops[op_index].size = size;
                max_index = (index > max_index) ? index : max_index;
                break;
            case 'f':
                fscanf(stdin, "%ud", &index);
                trace->ops[op_index].type = FREE;
                trace->ops[op_index].index = index;
                break;
            default:
                printf("Bogus type character (%c) in tracefile %s\n", type[0],
                       trace->filename);
                exit(1);
        }
        op_index++;
        if (op_index == trace->num_ops) break;
    }
    // fclose(tracefile);
    assert(max_index == trace->num_ids - 1);
    assert(trace->num_ops == op_index);
    strcpy(stats->filename, trace->filename);
    stats->weight = trace->weight;
    stats->ops = trace->num_ops;

    return trace;
}

static void reinit_trace(trace_t *trace) {
    memset(trace->blocks, 0, trace->num_ids * sizeof(*trace->blocks));
    memset(trace->block_sizes, 0, trace->num_ids * sizeof(*trace->block_sizes));
}

static void free_trace(trace_t *trace) {
    free(trace->ops);
    free(trace->blocks);
    free(trace->block_sizes);
    free(trace->block_rand_base);
    free(trace);
}

static int eval_mm_valid(trace_t *trace, range_t **g_ranges) {
    int i;
    int index;
    int size;
    //    int oldsize;
    char *newp;
    char *oldp;
    char *p;

    mem_reset_brk();
    clear_ranges(g_ranges);
    reinit_trace(trace);

    if (my_init() < 0) {
        // malloc_error(trace, 0, "my_init failed.");
        return 0;
    }
    for (i = 0; i < trace->num_ops; i++) {
        index = trace->ops[i].index;
        size = trace->ops[i].size;

        switch (trace->ops[i].type) {
            case ALLOC:
                if ((p = my_malloc((size_t)size)) == NULL) {
                    // malloc_error(trace, i, "my_malloc failed.");
                    return 0;
                }
                if (add_range(g_ranges, p, size, trace, i, index) == 0)
                    return 0;
                //                memset(p, index & 0xFF, size);
                trace->blocks[index] = p;
                trace->block_sizes[index] = (size_t)size;

                randomize_block(trace, index);

                break;
            case REALLOC:
                //                check_index(trace,i,index);

                oldp = trace->blocks[index];
                newp = my_realloc(oldp, size);
                if (newp == NULL && size != 0) {
                    // malloc_error(trace, i, "my_realloc failed.");
                    return 0;
                }

                if (newp != NULL && size == 0) {
                    // malloc_error(trace, i,
                    //              "my_realloc with size 0 returned
                    //              non-NULL.");
                    return 0;
                }

                remove_range(g_ranges, oldp);

                if (size > 0) {
                    if (add_range(g_ranges, newp, size, trace, i, index) == 0)
                        return 0;
                }

                trace->blocks[index] = newp;
                if (size < trace->block_sizes[index]) {
                    trace->block_sizes[index] = size;
                }
                if (check_index(trace, i, index) == 0) {
                    trace->block_sizes[index] = size;
                    randomize_block(trace, index);
                } else {
                    // malloc_error(trace, i,
                    //              "my_realloc did not preserve the "
                    //              "data from old block");
                    return 0;
                }

                break;

            case FREE:
                if (check_index(trace, i, index) != 0) {
                    // malloc_error(trace, i, "check index in free failed");
                    return 0;
                }

                if (index == -1) {
                    p = 0;
                } else {
                    p = trace->blocks[index];
                    remove_range(g_ranges, p);
                }

                my_free(p);
                break;

            default:
                app_error("Nonexistent request type in eval_mm_valid");
        }
    }
    return 1;
}

static double eval_mm_util(trace_t *trace, int tracenum) {
    int i;
    int index;
    int size, newsize, oldsize;
    int max_total_size = 0;
    int total_size = 0;
    char *p;
    char *newp, *oldp;

    reinit_trace(trace);
    mem_reset_brk();
    if (my_init() < 0) {
        // app_error("my_init failed in eval_mm_util");
        error++;
        return 0;
    }
    for (i = 0; i < trace->num_ops; i++) {
        switch (trace->ops[i].type) {
            case ALLOC:
                index = trace->ops[i].index;
                size = trace->ops[i].size;

                if ((p = my_malloc(size)) == NULL) {
                    // app_error("my_malloc failed in eval_mm_util");
                    error++;
                    return 0;
                }
                trace->blocks[index] = p;
                trace->block_sizes[index] = (size_t)size;
                total_size += size;
                max_total_size =
                    (total_size > max_total_size) ? total_size : max_total_size;
                break;

            case REALLOC:
                index = trace->ops[i].index;
                newsize = trace->ops[i].size;
                oldsize = (int)trace->block_sizes[index];

                oldp = trace->blocks[index];
                if ((newp = my_realloc(oldp, newsize)) == NULL &&
                    newsize != 0) {
                    // app_error("my_realloc failed in eval_mm_util");
                    error++;
                    return 0;
                }

                trace->blocks[index] = newp;
                trace->block_sizes[index] = (size_t)newsize;
                total_size += (newsize - oldsize);
                max_total_size =
                    (total_size > max_total_size) ? total_size : max_total_size;
                break;
            case FREE:
                index = trace->ops[i].index;
                if (index < 0) {
                    size = 0;
                    p = 0;
                } else {
                    size = (int)trace->block_sizes[index];
                    p = trace->blocks[index];
                }
                my_free(p);
                total_size -= size;
                break;
            default:
                app_error("Nonexistent request type in eval_mm_util");
        }
    }
    return ((double)max_total_size / (double)mem_heapsize());
}

static void eval_mm_speed(void *ptr) {
    int i, index, size, newsize;
    char *p, *newp, *oldp, *block;
    trace_t *trace = ((speed_t *)ptr)->trace;

    reinit_trace(trace);
    mem_reset_brk();
    if (my_init() < 0) {
        // app_error("my_init failed in eval_mm_util");
        error++;
        return;
    }
    for (i = 0; i < trace->num_ops; i++) switch (trace->ops[i].type) {
            case ALLOC:
                index = trace->ops[i].index;
                size = trace->ops[i].size;
                if ((p = my_malloc(size)) == NULL) {
                    // app_error("mm_malloc error in eval_mm_speed");
                    error++;
                    return;
                }
                trace->blocks[index] = p;
                break;

            case REALLOC:
                index = trace->ops[i].index;
                newsize = trace->ops[i].size;
                oldp = trace->blocks[index];
                if ((newp = my_realloc(oldp, newsize)) == NULL &&
                    newsize != 0) {
                    // app_error("mm_realloc error in eval_mm_speed");
                    error++;
                    return;
                }
                trace->blocks[index] = newp;
                break;

            case FREE:
                index = trace->ops[i].index;
                if (index < 0) {
                    block = 0;
                } else {
                    block = trace->blocks[index];
                }
                my_free(block);
                break;

            default:
                app_error("Nonexistent request type in eval_mm_speed");
        }
}

void app_error(const char *msg) {
    printf("%s\n", msg);
    exit(1);
}

void unix_error(const char *msg) {
    printf("%s: %s\n", msg, strerror(errno));
    exit(1);
}

// void malloc_error(const trace_t *trace, int opnum, char *msg) {
//     errors++;
//     printf("ERROR [trace %s, line %d]: %s\n", trace->filename,
//     LINENUM(opnum),
//            msg);
// }

File: topic1/solutiont11/malloc/code/judge/java/Tester.java
import java.io.*;
import java.util.*;

public class Tester {

    private Solve solver;
    private int verbose;
    private int error;
    private String flag;
    Tester() {
        error=0;
        verbose = 0;
        solver = new Solve();
        flag="77796755c34b3aada95b7ab2caf33d5ec337d567e58a328cf02c139b29bc650c";
    }

    final int lineNum(int i) {
        return i + 5;
    }

    final boolean isAligned(int pos) {
        return (pos % Config.ALIGNMENT) == 0;
    }


    class RangeList {

        class Range {

            int low;
            int high;
            Range next;

            public Range() {
                next = null;
            }

            public Range(int _low, int _high) {
                low = _low;
                high = _high;
                next = null;
            }

            public Range(int _low, int _high, Range _next) {
                low = _low;
                high = _high;
                next = _next;
            }
        }

        Range head;
        int size;

        public RangeList() {
            head = new Range();
            size = 0;
        }

        public boolean add(int low, int bsize, String filename, int opnum) {
            if (check(low, bsize, filename, opnum)) {
                Range tmp = new Range(low, low + bsize - 1);
                tmp.next = head.next;
                head.next = tmp;
                size++;

                return true;
            }
            return false;
        }

        private boolean check(int low, int bsize, String filename, int opnum) {
            int high = low + bsize - 1;
            if (!isAligned(low)) {                
                return false;
            }

            if ((low < solver.myUtils.memHeapLow()) || (low > solver.myUtils.memHeapHigh()) ||
                    (high < solver.myUtils.memHeapLow()) || (high > solver.myUtils.memHeapHigh())) {
                return false;
            }


            for (Range temp = head.next; null != temp; temp = temp.next) {
                if ((low >= temp.low && low <= temp.high) ||
                        (high >= temp.low && high <= temp.high)) {
                    
                    return false;
                }
            }
            return true;
        }


        public void remove(int low) {
            Range prev = head;
            Range current = head.next;

            while (current != null && current.low != low) {
                prev = current;
                current = current.next;
            }
            if(current==null)
                return;
            prev.next = current.next;
            size--;
        }

        public void clear() {
            head.next = null;
            size = 0;
        }

        public void printList() {
            if (size == 0)
                System.out.println("list null");
            else {
                for (Range cur = head.next; null != cur; cur = cur.next) {
                    System.out.println(cur.low + " " + cur.high);
                }
            }
        }

    }


    class TraceOp {
        public int type;
        public int index;
        public int size;

        public TraceOp(int _type, int _index, int _size) {
            type = _type;
            index = _index;
            size = _size;
        }

        void print() {
            System.out.printf("%d\t%d\t%d\n", type, index, size);
        }
    }

    class Trace {
        int numIds;
        int numOps;
        TraceOp[] ops;
        int[] blocksPos;
        int[] blockSize;
        String fileName;
        int weight;

        public Trace() {

        }

        public Trace(int _numIds, int _numOps, String _fileName, int _weight) {
            numIds = _numIds;
            numOps = _numOps;
            ops = new TraceOp[_numOps];
            blockSize = new int[_numIds];
            blocksPos = new int[_numIds];
            fileName = _fileName;
            weight = _weight;
        }

        void print() {
            System.out.println("trace print start");
            System.out.println(fileName);
            System.out.println(numIds);
            System.out.println(numOps);
            for (int i = 0; i < numOps; i++) {
                ops[i].print();
            }
            System.out.println("trace print end");
        }
    }


    class Stats {
        String fileName;
        double ops;
        boolean valid;
        double secs;
        double util;
        int weight;
    }



    void run(String[] args) {
        String dataDir = Config.DATADIR;
        String[] dataFiles = Config.DEFAULT_DATAFILES;

        int numFiles = Config.FILENUM;
        int errors = 0;
        int c;
        boolean onetime = false;

        numFiles = 1;
        dataDir = "./";
        dataFiles = new String[1];
        
        dataFiles[0] = "asdasd";

        Trace trace;
        Stats[] g_stats = new Stats[numFiles]; 


        RangeList g_rangeList = new RangeList();
        solver.myUtils.memInit();

        for (int i = 0; i < numFiles; ++i) {
            
            trace = readTrace(dataDir, dataFiles[i]);
            g_stats[i] = new Stats();
            g_stats[i].fileName = trace.fileName;
            g_stats[i].ops = trace.numOps;
            g_stats[i].weight = trace.weight;
            
            

            if (verbose > 1)
                System.out.printf("Checking for correctness, ");

            g_stats[i].valid = evalValid(trace, i, g_rangeList);

            if (onetime) {
                break;
            }
            if (g_stats[i].valid) {
                if (verbose > 1)
                    System.out.printf("utilization, ");

                g_stats[i].util = evalUtil(trace, i);

                if (verbose > 1)
                    System.out.printf("and time.\n");
                g_stats[i].secs = evalSpeed(trace, i, 10);
            }

            if (g_stats[i].valid && error == 0) {
                System.out.printf("%s %d %.6f %.6f\n", flag,1, 1 - g_stats[i].util,
                       g_stats[i].secs);
            } else {
                System.out.printf("%s %d %.6f %.6f\n", flag,0, 1.0,99999999.0);
            }
        }

        
        
        
        
        
        
        
        
        
        
        
        

        

        solver.myUtils.memDeInit();

    }

    private Trace readTrace(String traceDir, String fileName) {
        try {

            if (verbose > 1) {
                System.out.printf("Read testfile: %s\n", fileName);
            }
            
            

            
            
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));



            int i = 0;
            int weight = Integer.parseInt(br.readLine());
            int ids = Integer.parseInt(br.readLine());
            int ops = Integer.parseInt(br.readLine());
            int ignoteFlag = Integer.parseInt(br.readLine());

            if (weight < 0 || weight > 1) {
                
                System.exit(0);
            }
            if (ignoteFlag != 0 && ignoteFlag != 1) {
                
                System.exit(0);
            }

            Trace _trace = new Trace(ids, ops, fileName, weight);

            
            

            int opIndex = 0;
            int maxIndex = 0;
            int index = 0;
            String line = null;
            int size=0;


            while ((line = br.readLine()) != null) {
                String[] tmp = line.split(" ");
        
                index = Integer.parseInt(tmp[1]);
                if(tmp.length==3){
                    size=Integer.parseInt(tmp[2]);
                }
                switch (tmp[0]) {
                    case "a":
                        _trace.ops[opIndex] = new TraceOp(0, index, size);
                        maxIndex = (index > maxIndex) ? index : maxIndex;

                        break;
                    case "f":
                        _trace.ops[opIndex] = new TraceOp(1, index, 0);
                        break;
                    case "r":
                        _trace.ops[opIndex] = new TraceOp(2, index, size);
                        maxIndex = (index > maxIndex) ? index : maxIndex;
                        break;
                    default:
                        System.out.printf("Bogus type character (%s) in tracefile %s\n", tmp[0], fileName);
                        break;
                }
                opIndex++;
                if (opIndex == _trace.numOps) break;
            }
            br.close();

            
            if (maxIndex != _trace.numIds - 1) {
                System.out.println(maxIndex);
                System.out.println("maxIndex!=_trace.numIds");
                System.exit(0);
            }
            if (opIndex != _trace.numOps) {
                System.out.println("opIndex!=_trace.numOps");
                System.exit(0);
            }
            
            return _trace;

        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    private void reinitTrace(Trace trace){
        Arrays.fill(trace.blocksPos, -1);
    }

    private boolean evalValid(Trace trace, int tracenum, RangeList list) {


        
        
        solver.myUtils.memResetBrk();
        list.clear();
        reinitTrace(trace);

        if (solver.myInit() < 0) {
            
            return false;
        }

        int i, index, size, oldSize;
        int pos, oldPos;
        for (i = 0; i < trace.numOps; ++i) {
            index = trace.ops[i].index;
            size = trace.ops[i].size;
            switch (trace.ops[i].type) {
                case 0:
                    if ((pos = solver.myMalloc(size)) < 0) {
                        
                        return false;
                    }
                    if (!list.add(pos, size, trace.fileName, i))
                        return false;

                    for (int k = 0; k < size; k++) {
                        solver.myUtils.g_heap[pos + k] = (byte) (index & 0xff);
                        
                    }

                    trace.blocksPos[index] = pos;
                    trace.blockSize[index] = size;
                    break;
                case 1:
                    if (index == -1) {
                        pos = -1;
                    } else {
                        pos = trace.blocksPos[index];
                        list.remove(pos);
                    }
                    solver.myFree(pos);
                    break;

                case 2:
                    oldPos = trace.blocksPos[index];
                    pos = solver.myRealloc(oldPos, size);
                    if (pos < 0 && size != 0) {
                        
                        return false;
                    }

                    if (pos >= 0 && size == 0) {
                        
                        return false;
                    }

                    list.remove(oldPos);

                    if (size > 0) {
                        if (!list.add(pos, size, trace.fileName, i))
                            return false;
                    }

                    oldSize = trace.blockSize[index];
                    if (size < oldSize)
                        oldSize = size;

                    for (int j = 0; j < oldSize; j++) {
                        if (solver.myUtils.g_heap[pos + j] != ((byte) (index & 0xFF))) {
                            
                            return false;
                        }
                    }

                    for (int k = 0; k < size; k++) {
                        solver.myUtils.g_heap[pos + k] = (byte) (index & 0xff);
                    }

                    trace.blocksPos[index] = pos;
                    trace.blockSize[index] = size;
                    break;


                default:
                    appError("Nonexistent request type in evalValid");
                    break;
            }
        }

        return true;
    }

    private double evalUtil(Trace trace, int tracenum) {

        reinitTrace(trace);

        solver.myUtils.memResetBrk();

        if (solver.myInit() < 0) {
            
            error++;
            return -1;
        }

        int i, index, size, oldSize;
        int totalSize = 0;
        int maxTotalSize = 0;
        int pos, oldPos;
        for (i = 0; i < trace.numOps; ++i) {
            index = trace.ops[i].index;
            size = trace.ops[i].size;
            switch (trace.ops[i].type) {
                case 0:
                    if ((pos = solver.myMalloc(size)) < 0) {
                        
                        error++;

                        return -1;
                    }

                    trace.blocksPos[index] = pos;
                    trace.blockSize[index] = size;
                    totalSize += size;
                    maxTotalSize = Math.max(maxTotalSize, totalSize);
                    break;

                case 1:
                    if (index < 0) {
                        size = 0;
                        pos = -1;
                    } else {
                        pos = trace.blocksPos[index];
                        size = trace.blockSize[index];
                    }
                    solver.myFree(pos);
                    totalSize -= size;
                    break;

                case 2:
                    oldPos = trace.blocksPos[index];
                    if ((pos = solver.myRealloc(oldPos, size)) < 0 && size != 0) {
                        
                        error++;

                        return -1;
                    }

                    oldSize = trace.blockSize[index];

                    trace.blocksPos[index] = pos;
                    trace.blockSize[index] = size;
                    totalSize += (size - oldSize);
                    maxTotalSize = Math.max(maxTotalSize, totalSize);
                    break;

                default:
                    appError("Nonexistent request type in evalUtil");
                    break;
            }
        }


        return (double) maxTotalSize / (double) solver.myUtils.memHeapSize();
    }

    private void evalSpeedFunc(Trace trace, int tracenum) {
        reinitTrace(trace);
        solver.myUtils.memResetBrk();
        if (solver.myInit() < 0) {
            
            return;
        }
        int i, index, size;
        int pos, oldPos;
        for (i = 0; i < trace.numOps; ++i) {
            index = trace.ops[i].index;
            size = trace.ops[i].size;
            switch (trace.ops[i].type) {
                case 0:
                    if ((pos = solver.myMalloc(size)) < 0) {
                        
                        error++;

                        return;
                    }
                    trace.blocksPos[index] = pos;
                    break;

                case 1:
                    if (index < 0) {
                        pos = -1;
                    } else {
                        pos = trace.blocksPos[index];
                    }
                    solver.myFree(pos);
                    break;

                case 2:
                    oldPos = trace.blocksPos[index];
                    if ((pos = solver.myRealloc(oldPos, size)) < 0 && size != 0) {
                        
                        error++;

                        return;
                    }
                    trace.blocksPos[index] = pos;
                    break;
                default:
                    appError("Nonexistent request type in evalSpeedFunc");
                    break;
            }
        }
    }

    private double evalSpeed(Trace trace, int tracenum, int n) {
        long start = System.nanoTime();
        for (int i = 0; i < n; ++i) {
            evalSpeedFunc(trace, tracenum);
        }
        long end = System.nanoTime();
        double diff = (double) (end - start) / 1000000000L;
        return diff / n;
    }


    
    
    
    
    


    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    

    
    
    
    
    
    
    
    
    

    

    
    
    
    
    
    
    
    

    private void print(String msg) {
        System.out.println(msg);
    }

    private void appError(String msg) {
        print(msg);
        System.exit(0);
    }

    private void unixError(String msg) {
        print(msg);
        System.exit(0);
    }

    private void mallocError(String filename, int opnum, String msg) {
        System.out.format("ERROR [trace %s, line %d]: %s\n", filename, lineNum(opnum), msg);
    }


}
